src/typechecker.erl

```diff
@@ -2906,6 +2906,12 @@ type_check_fun(_Env, {remote, _, _Expr, _}, Arity)->
            lists:duplicate(Arity, type(any))},
           {type,0,any,[]}]},
         []]}], #{}, constraints:empty()};
+type_check_fun(Env, Var = {var, _, _Fun}, Arity) ->
+    case do_type_check_expr(Env, Var) of
+        FunType = {{type,0,'fun', [{type,0,product,Vars}, {type,0,any,[]}]}
+                 ,_,_} when length(Vars) =:= Arity -> FunType;
+        _ -> type_check_expr(Env, Var)
+    end;
```

Comment:
I think we shouldn't report any type error for `fun_wrong_arity/1`.

```Erlang
-spec fun_wrong_arity(any()) -> boolean().
fun_wrong_arity(Fun) when is_function(Fun, 2) ->
    Fun(0);
fun_wrong_arity(_Fun) -> true.
```

Any() is the dynamic type, so we don't want static type errors for it. The same if this function would have no spec at all.

(Unless you've found out: `type_check_fun` (and `type_check_expr`, etc., the ones without `_in`) are used for checking code against no particular type, propagating types outward, types originating from calls to functions with specs. The ones with `_in` are checking code against a particular type.)

Reply:
Ok, that's first of all a distinction I wasn't aware of, and for which the function names don't help a bit :no_mouth:

But in this function, the guard is a refinement. I have no idea what kind of evil things I'm giving to this function, but if it enters the first branch, then I do have information about it: it is a function of arity exactly two. That's all I know, so I cannot check the type of the arguments or the returns, but I can check its top-level type and its arity.

As a matter of fact, I believe the BEAM is a strongly typed machine, just that Erlang is dynamically checked, and not statically. And that is what this guard is doing, a dynamic check, right?

Reply:
Josef had this idea that only the specs are used as static type information, to have a clear rule for when there are type errors and when not. But now we're also using type specifiers in bitstring constructions (e.g. float in `<<(X + Y)/float>>`). Maybe it makes sense to also use this kind of guards as static type information. Or maybe not.

@josefs what's your thoughts on this?

Reply:
Yes, I think the guard adds a dynamic check. The compiler adds a lot of dynamic checks, but then in the optimization passes, some of the checks are removed if it can be proven that it's covered by an earlier check such as these guards.

Anyway, if the spec of the function would be `-spec fun_wrong_arity(fun((_,_) -> _) | fun((_) -> _)) -> boolean()`, this refinement is definitely wanted. I'm just not so sure about any() and that is all about allowing a dynamic style and not giving any errors in such cases. With the `--infer` flag, we actually do give type errors even if there are no specs. That's sort of why that flag is not enabled by default.

Reply:
Refining `any()` is clearly wrong to me, especially if the programmer has written the type as in your example @zuiderkwast.  That would defeat the purpose of the type.

But the idea of using guards as a source of type information is quite nice, in particular if inference is enabled.

Reply:
Right, so we can throw it all into type inference then. So a priority would be to have a way to test it, for example the `-gradualizer({infer, true}).` that we mentioned elsewhere. Then it's going to be tricky to distinguish in code if I want to refine `-spec fun_wrong_arity(fun((_,_) -> _) | fun((_) -> _)) -> boolean()` but not `-spec fun_wrong_arity(any()) -> boolean()`.

Then I'll just move all this code under that infer flag.

Uh, and is there anybody planning to refactor `typechecker.erl` a bit? :sob:

Reply:
I'm not convinced that refining `any()` is the wrong thing to do in general (for Gradualizer it might be). This TypeScript code errors since g is refined from any to string. I find it to be very similar to our guard situation:
```TypeScript
function f(g: any): boolean {
    if (typeof g === "string") {
        return g
    }
    return true
}
```

Reply:
any() means "something unknown between none() and everything". Consider this:

```Erlang
%% no spec
my_sleep(N) when is_integer(N) -> timer:sleep(N);
my_sleep(true) -> timer:sleep(1000);
my_sleep(false) -> ok.
```

If any() is refined to integer() in the first clause, a type error would be reported for calling timer:sleep/1, which requires non_negative_integer() | infinity. In this case, we can imagine the programmer knows that "of course it's a positive integer or a boolean", so this error is clearly not wanted.

In order to not fail for the above example, we can only refine it to *something unknown between none() and integer()*, i.e. we can infer an upper bound only. Perhaps we can be represent this using an intersection type `any() & integer()` (assuming & means intersection).

As a comparison, normalizing the *union* integer() | any() doesn't remove integer() from the type. This is because, when I implemented normalization of unions, I had some sort of intuition that this should represent *something unknown between integer() and everything*, i.e. any() with integer() as a lower bound.

This is what I think we should have with `infer` (so we can report an error for `fun_wrong_arity/1` but no error for `my_sleep/1`). Without infer, just any().

Reply:
@NelsonVides, regarding not refining any() while refining other types shouldn't be that tricky. Any() is handled differently from other things almost everywhere in the code.

> Then I'll just move all this code under that infer flag.

Perfect.

> Uh, and is there anybody planning to refactor typechecker.erl a bit?

Yes, it's in the back of my head. Renaming the functions `type_check_xxx(_in)`, `add_type(s)_pat(s)` and `add_any_type(s)_pat(s)` is a first step. Not sure what to name them though. :-) Ideas are welcome; perhaps open an issue to keep discussions separate?

Reply:
@Zalastax we should avoid using typescript as reference. That typesystem is unsound.

Reply:
@zuiderkwast please open a ticket to discuss refactoring of `typechecker.erl` and post your ideas there.

Reply:
@zuiderkwast I wonder whether there is any way that you can make your intuition about intersection and union types a little more formal? I know you may not have that much experience with formalizing type systems but it'd be nice to try. I'm happy to help, if you can give a little bit more detail about your thinking.

The reason I ask is because your ideas sound compelling. They are also different from the existing literature: https://www.irif.fr/~gc/papers/icfp17.pdf which makes them extra intriguing. And I don't think your ideas are necessarily wrong just because they're different from a published paper.

Let me just give a tldr of the paper: `any() & integer()` is the same as `integer()` and `any() | integer()` is `any()`.

Reply:
I think this paper is saying exactly what I am trying to say. This is spot on! Thanks @josefs for finding it! And no, `any() & integer()` is not the same as `integer()`.

Here, from the paper, the bottom of page 2, an example with type casts (AKA type assertions):

> ```ML
> let f (condition : Bool) (x : (Int | Bool)) : (Int | Bool) =
>   if condition then
>     succ ((Int) x)
>   else
>     not ((Bool) x)
> ```
>
> The code above shows that the use of union types fixes the shortcoming we pointed out. However in order to ensure that the applications of `succ` and `not` are both well typed we also had to add two type casts that check at run-time whether the argument has the required type and raise an exception otherwise.

Then, on the next page:

> By using set-theoretic types in conjunction with gradual types, it is possible both to ensure that `f` will only be applied to booleans or integers *and* to delegate the insertion of type casts to the system. This is shown by the following piece of code that our system will compile into the previous one.
>
> ```ML
> let f (condition : Bool) (x : (Int | Bool) & ?) : (Int | Bool) =
>   if condition then
>     succ x
>   else
>     not x
> ```
>
> In this example, the variable is of type `((Int | Bool) & ?)`, where “&” denotes an intersection. This indicates that x has both type `(Int | Bool)` and type `?`. Intuitively, this means that the function `f` accepts as a second argument a value of any type (which is indicated by `?`), as long as this value is also an integer *or* a Boolean. The effect of having added “& ?” in the type of the parameter is that the programmer is no longer required to add the explicit casts in the body of the function: the system will take care of it at compile time.

In other words, when adding "& any()" to a type, the type assertions can be removed. The type checker doesn't report a type error.

The explanation continues in the paper:

> The combination of a union type with “& ?” to type a parameter corresponds to a programming style in which the programmer asks the system to *statically* enforce that the function will be applied only to arguments in the union type and delegates to the system the *dynamic* checking, where/if necessary, [...]

Isn't this spot on?

Reply:
> Isn't this spot on?

No :-)

The part of the paper you're quoting has to do with how to compile the program in such a way that those parts which are statically typed are compiled without runtime checks. This is called "sound gradual typing" and means that whenever the execution switches from dynamically typed to statically typed, the code needs to first check that the values are indeed of the type that the static code expects them to be. The paper uses the word "cast" for this check. This part is tricky and that's what they're talking about in your quotes. 

In gradualizer we are not concerned with casts because after typechecking we throw away the types and compile the code as if it was completely dynamically typed.

Reply:
Correct me if I'm wrong below.

Let's say we annotate every expression with a type using erl_anno. If the type system assigns an unknown type to an expression, it delegates the type checking to the runtime, which means that the compiler needs to insert type checks whenever the type is unknown where a specific type is needed. If the type system instead assigns a static type to an expression, it can be trusted by the compiler. That's the gradual guarantee, right?

If we would tag the expressions with a static type and the compiler would actually make use of this type, our type assertions would actually have to make sure that the value is of the that type, e.g. by inserting code asserts the type in runtime and raises an exception if it isn't (e.g. `?assert(is_integer(X))`). Only then, the type assertion can soundly assert a static type, which the compiler can trust. The only reason we don't is that the compiler treats every type as unknown anyway and inserts runtime checks everywhere.

I can't see how the rules for intersection and union should be different just because the current compiler doesn't make use of our type information.